source/common/upstream/upstream_impl.cc

```diff
@@ -271,132 +198,160 @@ UpstreamLocalAddressSelectorImpl::buildBaseSocketOptions(
-    }
-
-    upstream_local_addresses_.push_back(extra_upstream_local_address);
-  }
-
-  if (bind_config.additional_source_addresses_size() == 1) {
-    UpstreamLocalAddress additional_upstream_local_address;
-    additional_upstream_local_address.address_ =
-        Network::Address::resolveProtoSocketAddress(bind_config.additional_source_addresses(0));
-    ASSERT(additional_upstream_local_address.address_->ip() != nullptr &&
-           upstream_local_address.address_->ip() != nullptr);
-    if (additional_upstream_local_address.address_->ip()->version() ==
-        upstream_local_address.address_->ip()->version()) {
+    if (!bind_config->has_source_address() &&
+        (bind_config->extra_source_addresses_size() > 0 ||
+         bind_config->additional_source_addresses_size() > 0)) {
       throw EnvoyException(fmt::format(
-          "{}'s upstream binding config has two same IP version source addresses. Only two "
-          "different IP version source addresses can be supported in BindConfig's source_address "
-          "and additional_source_addresses fields",
-          cluster_name.empty() ? "Bootstrap" : fmt::format("Cluster {}", cluster_name)));
+          "{}'s upstream binding config has extra/additional source addresses but no "
+          "source_address. Extra/additional addresses cannot be specified if "
+          "source_address is not set.",
+          !(cluster_name.has_value()) ? "Bootstrap"
+                                      : fmt::format("Cluster {}", cluster_name.value())));
     }
-
-    additional_upstream_local_address.socket_options_ =
-        std::make_shared<Network::ConnectionSocket::Options>();
-
-    Network::Socket::appendOptions(additional_upstream_local_address.socket_options_,
-                                   base_socket_options);
-    Network::Socket::appendOptions(additional_upstream_local_address.socket_options_,
-                                   cluster_socket_options);
-
-    upstream_local_addresses_.push_back(additional_upstream_local_address);
   }
+  UpstreamLocalAddressSelectorFactory* local_address_selector_factory;
+  const envoy::config::core::v3::TypedExtensionConfig* const local_address_selector_config =
+      bind_config.has_value() && bind_config->has_local_address_selector()
+          ? &bind_config->local_address_selector()
+          : nullptr;
+  if (local_address_selector_config) {
+    local_address_selector_factory =
+        Config::Utility::getAndCheckFactory<UpstreamLocalAddressSelectorFactory>(
+            *local_address_selector_config, false);
+  } else {
+    // Create the default local address selector if one was not specified.
+    envoy::extensions::upstream::local_address_selector::v3::DefaultLocalAddressSelector
+        default_config;
+    envoy::config::core::v3::TypedExtensionConfig typed_extension;
+    typed_extension.mutable_typed_config()->PackFrom(default_config);
+    local_address_selector_factory =
```

Comment:
This is a bit problematic as it requires the extension to be linked (hence, not an extension).
Can it just call the function that create the `UpstreamLocalAddressSelectorFactory`?

Reply:
I did have the default implementation defined in `//source/common/upstream:upstream_lib` and created it without going through the filter registry. I was asked in review comments to implement it this way.

What do you think is the best approach? @soulxu @adisuissa

Reply:
I thought you only reference the ext API here,  so you needn't build with the extension as @adisuissa pointed out at https://github.com/envoyproxy/envoy/pull/27705/files#r1279293248

So I think you can just remove the deps on the build, then it is ok?

Reply:
Take a look at [DefaultClientConnectionFactory](https://github.com/envoyproxy/envoy/blob/104e0635e894ad75e86f8cc0c52f7d4847fc6510/source/common/network/default_client_connection_factory.h#L16C7-L16C37) that implements [ClientConnectionFactory](https://github.com/envoyproxy/envoy/blob/104e0635e894ad75e86f8cc0c52f7d4847fc6510/envoy/network/client_connection_factory.h#L14), and the (non-default) extension [InternalClientConnectionFactory](https://github.com/envoyproxy/envoy/blob/104e0635e894ad75e86f8cc0c52f7d4847fc6510/source/extensions/bootstrap/internal_listener/client_connection_factory.h#L19).

Reply:
Thanks so much @adisuissa. I'll move the default implementation out of the extensions directory.

@soulxu I don't think that would work, as the default is chosen when nothing is specified in the config.

Reply:
thanks, I understand what you are saying here now